src/module.c

```diff
@@ -3540,6 +3540,241 @@ int RM_PublishMessageShard(RedisModuleCtx *ctx, RedisModuleString *channel, Redi
+    if (de == NULL) {
+        subscribers = listCreate();
+        dictAdd(channelModuleSubscribers, channelToAdd, subscribers);
+        incrRefCount(channelToAdd);
+    } else {
+        subscribers = dictGetVal(de);
+    }
+    listAddNodeTail(subscribers, ctx->module);
+
+    serverLog(LL_DEBUG,"Module %s subscribed to channel %s.", ctx->module->name, (char *)channelToAdd->ptr);
+
+    return;
+}
+
+/**
+ * This API unsubscribe(s) the module from the given channel.
+ * 
+ */
+void RM_UnsubscribeFromChannel(RedisModuleCtx *ctx, RedisModuleString *channel) {
+    dictEntry *de = dictFind(moduleChannelSubscribers, ctx->module);
+    if (de == NULL) {
+        return;
+    }
+
+    RedisModuleChannelSubscriber *moduleChannelSubscriber = dictGetVal(de);
+
+    /* No subscription found for module. */
+    if (listLength(moduleChannelSubscriber->subscriptions) == 0) {
+        return;
+    }
+
+    listNode *ln = listSearchKey(moduleChannelSubscriber->subscriptions, channel);
+
+    if (ln == NULL) {
+        return;
+    }
+
+    /* Find and remove the module subscriber to the channel. */
+    removeModuleFromChannelSubscription(ctx->module, ln->value);
+    listDelNode(moduleChannelSubscriber->subscriptions, ln);
+    if (listLength(moduleChannelSubscriber->subscriptions) == 0) {
+        cleanupModuleChannelSubscriber(ctx->module, moduleChannelSubscriber);
+    }
+    serverLog(LL_DEBUG,"Module %s unsubscribed from channel %s.", ctx->module->name, (char *)channel->ptr);
+}
+
+/**
+ * List all global channels a module is subscribed to.
+ * 
+ * Returns an array of RedisModuleString pointer in which the last entry is NULL. 
+ * The array must be freed by the caller via `RM_FreeGlobalChannelSubscriptionList`.
+ */
+RedisModuleString **RM_GlobalChannelSubscriptionList(RedisModuleCtx *ctx, size_t *numchannels) {
+    *numchannels = 0;
+    dictEntry *de = dictFind(moduleChannelSubscribers, ctx->module);
+    if (de == NULL) {
+        return NULL;
+    }
+
+    RedisModuleChannelSubscriber *moduleChannelSubscriber = dictGetVal(de);
+    listIter li;
+    listRewind(moduleChannelSubscriber->subscriptions, &li);
+
+    int channelCount = listLength(moduleChannelSubscriber->subscriptions);
+    robj **channels = zmalloc((channelCount + 1) * sizeof(robj*));
+
+    listNode *ln;
+    while ((ln = listNext(&li)) != NULL) {
+        robj* channel = listNodeValue(ln);
+        incrRefCount(channel);
+        channels[(*numchannels)++] = channel;
+    }
+    channels[*numchannels] = NULL;
+    return channels;
+}
+
+
+/**
+ * Cleanup up the global channels list response which was provided to the client via `GlobalChannelSubscriptionList`.
+ *
+ */
+void RM_FreeGlobalChannelSubscriptionList(RedisModuleString **channels) {
+    if (channels == NULL) return;
+    for (int i = 0; channels[i]; i++) {
+        decrRefCount(channels[i]);
+    }
+    zfree(channels);
+}
+
+/**
+ * Unsubscribe all the channel(s) this module had subscribed to.
+ * 
+ * @param module 
```

Comment:
since when did we start using this? for doxygen?

Reply:
CLion :|